packages/server/src/extensions/devicePublicKey/verifyDpkSignature.ts

```diff
@@ -0,0 +1,132 @@
+import base64url from "base64url";
+import { toHash } from "../../helpers/toHash";
+import { verifySignature } from "../../helpers/verifySignature";
+import { DevicePublicKeyAuthenticatorOutput } from "../../helpers/decodeAuthenticatorExtensions";
+import { SettingsService } from "../../services/settingsService";
+import { AttestationFormatVerifierOpts } from "../../registration/verifyRegistrationResponse";
+import { parseAuthenticatorData } from "../../helpers/parseAuthenticatorData";
+import { verifyAttestationFIDOU2F } from '../../registration/verifications/verifyAttestationFIDOU2F';
+import { verifyAttestationPacked } from '../../registration/verifications/verifyAttestationPacked';
+import { verifyAttestationAndroidSafetyNet } from '../../registration/verifications/verifyAttestationAndroidSafetyNet';
+import { verifyAttestationTPM } from '../../registration/verifications/tpm/verifyAttestationTPM';
+import { verifyAttestationAndroidKey } from '../../registration/verifications/verifyAttestationAndroidKey';
+import { verifyAttestationApple } from '../../registration/verifications/verifyAttestationApple';
+import {
+  AuthenticationCredentialJSON,
+  AuthenticatorAssertionResponseJSON,
+  RegistrationCredentialJSON,
+  AuthenticatorAttestationResponseJSON
+} from "@simplewebauthn/typescript-types";
+import { decodeAttestationObject } from "../../helpers/decodeAttestationObject";
+
+export type VerifyDpkSignatureOpts = {
+  credential: RegistrationCredentialJSON | AuthenticationCredentialJSON
+  devicePubKey: DevicePublicKeyAuthenticatorOutput;
+  signature: Buffer;
+};
+
+/**
+ * https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1663.html#sctn-device-publickey-extension-verification-create
+ * 3. Verify that `signature` is a valid signature over the assertion signature
+ *    input by the device public key *dpk*. (The signature algorithm is the same
+ *    as for the user credential.)
+ * @param options 
+ * @returns Promise<boolean>
+ */
+export async function verifyDpkSignature(
+  options: VerifyDpkSignatureOpts
+): Promise<boolean> {
+  const { credential, devicePubKey, signature } = options;
+
+  let authData;
+  if (isAuthenticationResponse(credential)) {
+    const { authenticatorData } = credential.response;
+    authData = base64url.toBuffer(authenticatorData);
+  }
+  if (isRegistrationResponse(credential)) {
+    const attestationObject = base64url.toBuffer(credential.response.attestationObject);
+    const { authData: authenticatorData } = decodeAttestationObject(attestationObject);
+    authData = authenticatorData;
+  }
```

Comment:
Authentications happen more often than registrations, so we can probably completely skip the `isRegistrationResponse()` the majority of the time with an `else if` here. Allowing both type guards to execute when we know one will always be false can probably be considered "inefficient":

```suggestion
  if (isAuthenticationResponse(credential)) {
    const { authenticatorData } = credential.response;
    authData = base64url.toBuffer(authenticatorData);
  } else if (isRegistrationResponse(credential)) {
    const attestationObject = base64url.toBuffer(credential.response.attestationObject);
    const { authData: authenticatorData } = decodeAttestationObject(attestationObject);
    authData = authenticatorData;
  }
```